src/porepy/numerics/ad/secondary_operator.py

```diff
@@ -0,0 +1,897 @@
+        """Mixed-dimensional domain passed at instantiation."""
+
+    def __call__(self, domains: pp.GridLikeSequence) -> pp.ad.Operator:
+        """Call this operator on a subset of grid in the md-grid to get a property
+        restricted to the passed domains.
+
+        1. If called on an empty list, a wrapped, empty array is returned.
+        2. If called on boundary grids, a time-dependent dense array is returned.
+        3. If called on subdomains on interfaces, a :class:`SecondaryOperator` is
+           returned.
+
+        Parameters:
+            domains: A subset of either grids, mortar grids or boundary grids in the
+                md-grid passed at instantiation.
+
+        Raises:
+            ValueError: If ``domains`` is not composed of either grids, mortar grids or
+                boundary grids.
+
+        """
+
+        # This is for completeness reasons, when calling equations on empty list
+        if len(domains) == 0:
+            return pp.wrap_as_dense_ad_array(np.zeros((0,)), name=self.name)
+        # On the boundary, this is a Time-Dependent dense array
+        elif all(isinstance(g, pp.BoundaryGrid) for g in domains):
+            return pp.ad.TimeDependentDenseArray(self.name, domains)
+        # On subdomains or interfaces, create the secondary operators
+        elif all(isinstance(g, pp.Grid) for g in domains) or all(
+            isinstance(g, pp.MortarGrid) for g in domains
+        ):
+            # for mypy
+            domains_ = cast(Sequence[pp.Grid] | Sequence[pp.MortarGrid], domains)
+            children = [child(domains_) for child in self._dependencies]
+
+            # Check if first-order dependency
+            assert all(isinstance(child, pp.ad.Variable) for child in children), (
+                "Secondary expressions must depend on independent variables, not"
+                + f" {[type(c) for c in children]}."
+            )
+
+            # always start with operator at current time step, current iterate
+            op = SecondaryOperator(
+                name=self.name,
+                domains=domains_,
+                children=children,
+            )
+
+            # assign the function which extracts the data
+            op.fetch_data = self.fetch_data
+
+            return op
+        else:
+            raise ValueError(
+                f"Unsupported domain configuration {[type(g) for g in domains]}."
+            )
+
+    def fetch_data(
+        self, op: SecondaryOperator, grid: pp.GridLike, get_derivatives: bool = False
+    ) -> np.ndarray:
+        """Function fetching the data stored for this secondary expression, represented
+        in Ad form by ``op``.
+
+        ``op`` has a time step or iterate index, which specifies which data should be
+        fetched.
+
+        This function is assigned to :attr:`SecondaryOperator.fetch_data`
+        (a work-around such that the Ad operator has no reference to the md-grid like
+        the other operators).
```

Comment:
Having read more code, I can guess what the answer will be, but I still think it will be useful to discuss this briefly, and perhaps document the answer as an implementation note for future reference.

Reply:
Having thought even more: Is this really a `_parse` method, like the other operators have?

Reply:
Discussed in person, documentation updated in the *Notes* sections in the module-level docstring.